### DISCUSSION TITLE
An error occurred when using MeshModifiers and Pressure

### DISCUSSION THREAD

**Discussion by PEI0214** [idaholab/moose#28946] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

When I used MeshModifiers, and set the Pressure boundary condition, I got the following error:
~~~
*** ERROR ***
'Pressure_load_0' of type 'Pressure' depends on variable(s) 'disp_x,disp_y,disp_z'. However, that variable does not appear to be defined on (all of) boundary 'top'.
~~~
My input file is ‘test1.i’ below. Is there a way to solve this problem?
<details>

<summary>test1.i</summary>

```
[Problem]
  kernel_coverage_check = false
  material_coverage_check = false
  boundary_restricted_node_integrity_check = false
[]

new_domain_blocks = '1001	1002'

[Mesh]
  [box]
    type = GeneratedMeshGenerator
    dim = 3
    nx = 10
    ny = 10
    nz = 10
    xmin = 0
    xmax = 1
    ymin = 0
    ymax = 1
    zmin = 0
    zmax = 1
  []
  [box1]
    type = SubdomainBoundingBoxGenerator
    input = box
    block_id = 1
    bottom_left = '0 0.5 0'
    top_right = '1 1 1'
  []

  add_subdomain_ids = '${new_domain_blocks}'
[]

[GlobalParams]
  displacements = 'disp_x disp_y disp_z'
[]

[Variables]
  [disp_x]
    block = '${new_domain_blocks}'
  []
  [disp_y]
    block = '${new_domain_blocks}'
  []
  [disp_z]
    block = '${new_domain_blocks}'
  []
[]

[Physics/SolidMechanics/QuasiStatic]
  [./all]
    strain = finite
    add_variables = false
    block = '${new_domain_blocks}'
  [../]
[]

[Kernels]
  [Gravity1]
    type = Gravity
    block = '${new_domain_blocks}'
    use_displaced_mesh = false
    variable = disp_y
    value = -10
  []
[]

[Functions]
  [zhouya]
    type = PiecewiseLinear
    x = '0  500 1000'
    y = '0  0   1e6'
  []
[]

# ===== Boundary Conditions =====
[BCs]
  [archor_x]
    type = DirichletBC
    boundary = 'bottom'
    variable = disp_x
    value = 0
  []

  [archor_y]
    type = DirichletBC
    boundary = 'bottom'
    variable = disp_y
    value = 0
  []

  [archor_z]
    type = DirichletBC
    boundary = 'bottom'
    variable = disp_z
    value = 0
  []

  [Pressure]
    [load]
      boundary = 'top'
      function = 'zhouya'
    []
  []
[]

# ===== Materials (linear-elastic to keep it simple) =====
[Materials]
  [elasticity_tensor1]
    type = ComputeIsotropicElasticityTensor
    block = '${new_domain_blocks}'
    youngs_modulus = 50E6 # 50 MPa
    poissons_ratio = 0.3
  []

  [stress]
    type = ComputeFiniteStrainElasticStress
    block = '${new_domain_blocks}'
  []

  [density1]
    type = GenericConstantMaterial
    block = '${new_domain_blocks}'
    prop_names = density
    prop_values = 2000
  []
[]

[MeshModifiers]
  [m1]
    type = TimedSubdomainModifier
    times = '200 300'
    blocks_from = '0	1'
    blocks_to = '1001	1002'
  []
[]

# ===== Executioner =====
[Executioner]
  type = Transient
  automatic_scaling = true

  end_time = 1000
  dt = 100

  solve_type = 'PJFNK'

  nl_abs_tol = 1E-5
  nl_max_its = 400

  l_tol = 1E-8
  l_max_its = 200
[]

[Outputs]
  exodus = true
[]

```

</details>

**Comment by GiudGiud** [comment]
Hello

So 'top' is not connected to the domains where the displacements are being solved for.
Is this intentional?

You could define the displacements everywhere on the domain, then use a NullKernel on the domains where you are not solving the solid mechanics equations for yet. 

Otherwise, you will need to turn off the other integrity check for nodal variables and BCs

Guillaume

**Reply by PEI0214** [reply]
>So 'top' is not connected to the domains where the displacements are being solved for. Is this intentional?

It wasn't my intention.

>You could define the displacements everywhere on the domain, then use a NullKernel on the domains where you are not solving the solid mechanics equations for yet.

I was trying to simulate the construction. If I define displacement on all domains and use NullKernel, the unconstructed regions will also be displaced.

>Otherwise, you will need to turn off the other integrity check for nodal variables and BCs

I tried to turn off 'boundary_restricted_elem_integrity_check', but it had the following error.
~~~
Time Step 0, time = 0

Time Step 1, time = 100, dt = 100

Performing automatic scaling calculation

Segmentation fault (core dumped)
~~~

**Reply by GiudGiud** [reply]
ok the integrity check is there for a reason. Consider this is not supported.

I would recommend trying the no-restriction + null kernel approach

**Reply by PEI0214** [reply]
If I define displacement on all domains and use NullKernel on subdomains that are not activated. Then the inactive subfields will also be deformed. As shown in the image below:
![image](https://github.com/user-attachments/assets/6bd55c1f-a026-4e30-992f-dfa66d4852ef)

This is also the case when setting DirichletBC on the 'bottom' boundary. The problem is resolved when 'boundary_restricted_node_integrity_check' is turned off.

**Reply by GiudGiud** [reply]
> The problem is resolved when 'boundary_restricted_node_integrity_check' is turned off.

This should not be. It s just a diagnostic to make sure the nodal BCs are applied on surfaces on which the variable is defined. 

> Then the inactive subfields will also be deformed

Are they deformed or just translated?
